products/libllbuild/BuildDB-C-API.cpp

```diff
@@ -0,0 +1,286 @@
+  }
+  
+  KeyID getKeyID(const KeyType &key) override {
+    if (!cAPIDelegate.get_key_id) {
+      return 0;
+    }
+    
+    return cAPIDelegate.get_key_id(cAPIDelegate.context, (char *)key.c_str());
+  }
+};
+  
+class CAPIBuildDB {
+  CAPIBuildDBDelegate cAPIDelegate;
+  
+  std::unique_ptr<BuildDB> _db;
+  
+public:
+  CAPIBuildDB(StringRef path, uint32_t clientSchemaVersion, CAPIBuildDBDelegate delegate, std::string *error_out): cAPIDelegate(delegate) {
+    _db = createSQLiteBuildDB(path, clientSchemaVersion, error_out);
+    _db.get()->attachDelegate(&cAPIDelegate);
+  }
+  
+  const uint64_t getCurrentIteration(bool *success_out, std::string *error_out) {
+    return _db.get()->getCurrentIteration(success_out, error_out);
+  }
+  
+  void setCurrentIteration(uint64_t value, std::string *error_out) {
+    _db.get()->setCurrentIteration(value, error_out);
+  }
+  
+  const bool lookupRuleResult(KeyID keyID, KeyType ruleKey, Result *result_out, std::string *error_out) {
+    return _db.get()->lookupRuleResult(keyID, ruleKey, result_out, error_out);
+  }
+  
+  const bool buildStarted(std::string *error_out) {
+    return _db.get()->buildStarted(error_out);
+  }
+  
+  void buildComplete() {
+    _db.get()->buildComplete();
+  }
+  
+  const bool getKeys(std::map<KeyID, KeyType>& keys_out, std::string *error_out) {
+    return _db.get()->getKeys(keys_out, error_out);
+  }
+  
+  void dump() {
+    _db.get()->dump(llvm::outs());
+  }
+};
+
+}
+
+const llb_data_t mapData(std::vector<uint8_t> input) {
+  const auto data = input.data();
+  const auto size = sizeof(data) * input.size();
+  const auto newData = (uint8_t *)malloc(size);
+  memcpy(newData, data, size);
+  
+  return llb_data_t {
+    input.size(),
+    newData
+  };
+}
+
+const llb_database_result_t mapResult(Result result) {
+  
+  auto count = result.dependencies.size();
+  auto data = result.dependencies.data();
+  auto size = sizeof(data) * count;
+  KeyID *deps = (KeyID*)malloc(size);
+  memcpy(deps, data, size);
+  
+  return llb_database_result_t {
+    mapData(result.value),
+    result.signature.value,
+    result.computedAt,
+    result.builtAt,
+    deps,
+    static_cast<uint32_t>(count)
+  };
+}
+
+void llb_database_destroy_result(llb_database_result_t *result) {
+  delete result->dependencies;
+}
+
+const llb_database_t* llb_database_create(
+                                    char *path,
+                                    uint32_t clientSchemaVersion,
+                                    llb_database_delegate_t delegate,
+                                    llb_data_t *error_out) {
+  assert(delegate.get_key_for_id);
+  assert(delegate.get_key_id);
+  
+  std::string error;
+  
+  auto database = new CAPIBuildDB(StringRef(path), clientSchemaVersion, delegate, &error);
+  
+  if (!error.empty() && error_out) {
+    error_out->length = error.size();
+    error_out->data = (const uint8_t*)strdup(error.c_str());
+  }
+  
+  return (llb_database_t *)database;
+}
```

Comment:
I would recommend starting with an initial PR which just proposes the `db.h`, and the basic database create and destroy methods. Then we can iterate individually on the additional APIs.

The general direction looks roughly right though.